nixos/modules/services/networking/pangolin.nix

```
@@ -0,0 +1,529 @@
+        type = with lib.types; nullOr path;
+        default = null;
+        description = ''
+          Path to a file containing sensitive environment variables for Pangolin. See https://docs.fossorial.io/Pangolin/Configuration/config for more information.
+          These will overwrite anything defined in the config.
+          The file should contain environment-variable assignments like:
+          ```
+          SERVER_SECRET=1234567890abc
+          ```
+        '';
+        example = "/etc/nixos/secrets/pangolin.env";
+      };
+
+      gerbilPort = lib.mkOption {
+        type = lib.types.port;
+        default = 3003;
+        description = ''
+          Specifies the port to listen on for Gerbil.
+        '';
+      };
+
+      gerbilEnvironmentFile = lib.mkOption {
+        type = nullOrOpt lib.types.path;
+        default = null;
+        description = ''
+          Path to a file containing sensitive environment variables for Gerbil. See https://docs.fossorial.io/Pangolin/Configuration/config for more information.
+          These will overwrite anything defined in the config.
+        '';
+        example = "/etc/nixos/secrets/gerbil.env";
+      };
+
+      traefikEnvironmentFile = lib.mkOption {
+        type = nullOrOpt lib.types.path;
+        default = null;
+        description = ''
+          Path to a file containing sensitive environment variables for Traefik. See https://docs.fossorial.io/Pangolin/Configuration/config for more information.
+          These will overwrite anything defined in the config.
+        '';
+        example = "/etc/nixos/secrets/traefik.env";
+      };
+
+      dataDir = lib.mkOption {
+        type = lib.types.str;
+        default = "/var/lib/pangolin";
+        example = "/srv/pangolin";
+        description = "Path to variable state data directory for Pangolin.";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    assertions =
+      (lib.mapAttrsToList (name: value: {
+        # check if the value is optional by looking at the type
+        assertion = (value == null) -> options.services.pangolin."${name}".type._optional or false;
+        message = "services.pangolin.${name} must be provided when Pangolin is enabled.";
+      }) cfg)
+      ++ [
+        {
+          # wildcards implies (dnsProvider and traefikEnvironmentFile)
+          assertion =
+            finalSettings.domains.domain1.prefer_wildcard_cert
+            -> (cfg.dnsProvider != "" && cfg.traefikEnvironmentFile != null);
+          message = "services.pangolin.dnsProvider and traefikEnvironmentFile must be provided when prefer_wildcard_cert is true.";
+        }
+      ];
+
+    networking.firewall = lib.mkIf cfg.openFirewall {
+      allowedTCPPorts = [
+        80
+        443
+      ];
+      allowedUDPPorts = [ 51820 ];
+    };
+
+    users = {
+      users = {
+        pangolin = {
+          description = "Pangolin service user";
+          group = "fossorial";
+          isSystemUser = true;
```

```suggestion
          isSystemUser = true;
          packages = [ cfg.package ];
```
Now that `pangctl` is a thing, adding the package to the Pangolin system user's environment would allow people to access it without changing file permissions or using the `root` user.
This would need to be added to the future manual entry. Something along the lines of "To change the administrator user credentials, use `run0 -u pangolin pangctl ...`"